File: src/1125-Smallest-Sufficient-Team/1125.go
import "math"
var people_skill []int
var skill_people [][]int
var smallest_team map[int][]int
func smallestSufficientTeam(req_skills []string, people [][]string) []int {
    key := make(map[string]int)
    for i, v := range req_skills {
        key[v] = i
    }
    people_skill = make([]int, len(people))
    skill_people = make([][]int, len(req_skills))
    for i, v := range people {
        for _, skill := range v {
            if _, ok := key[skill]; ok {
                people_skill[i] |= (1 << uint(key[skill]));
                skill_people[key[skill]] = append(skill_people[key[skill]], i)
            }
        }
    }
    smallest_team = map[int][]int{0:{}}
    return dfs((1 << uint(len(req_skills))) -1)
}

func dfs(skills_wanted int) []int {
    if v, ok := smallest_team[skills_wanted]; ok {
        return v
    }
    smallest_team_size := 61
    res := []int{}
    target_skill := int(math.Log2(float64(skills_wanted & -skills_wanted)))
    for _, p := range skill_people[target_skill] {
        cand_skill_wanted := skills_wanted &^ people_skill[p]
        pre := dfs(cand_skill_wanted)
        pre = append(pre, p)
        if len(pre) < smallest_team_size {
            smallest_team_size = len(pre)
            res = pre
        }
    }
    smallest_team[skills_wanted] = res
    return res
}

File: src/1125-Smallest-Sufficient-Team/1125.py
from functools import lru_cache
class Solution:
    def smallestSufficientTeam(self, req_skills: List[str], people: List[List[str]]) -> List[int]:
        key = {skill: i for i, skill in enumerate(req_skills)}
        people_skill = [0 for _ in range(len(people))] 
        skill_people = [[] for _ in range(len(req_skills))] 
        for i, p in enumerate(people):
            for skill in p:
                if skill in key:
                    people_skill[i] |= 1 << key[skill]
                    skill_people[key[skill]].append(i)
        
        def first_set_bit(n: int) -> int:
            return int(math.log2(n & -n)) #low bit
        
        @lru_cache()
        def dfs(skills_wanted: int) -> List[int]:
            if skills_wanted == 0:
                return []
            
            smallest_team_size = float('inf')
            res = []
            
            target_skill = first_set_bit(skills_wanted)
            
            for p in skill_people[target_skill]:
                cand_skills_wanted = skills_wanted
                cand_skills_wanted &= ~people_skill[p]
                cand = [p] + dfs(cand_skills_wanted)
                if len(cand) < smallest_team_size:
                    smallest_team_size = len(cand)
                    res = cand
            
            return res
        
        return dfs((1 << len(req_skills)) - 1)